add sdkch to maintain changelog effectively and free of file conflicts

alessio opened the pull request:
Imported from https://github.com/alessio/modcl

Closes: #2380

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [x] Added entries in `PENDING.md` with issue # 
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

rigelrozanski commented:
@alexanderbez I actually think this makes sense _not_ to include this tool in `cmd/` and also to never run it from a binary (but always with `go run` from the makefile) - I think of it more of a bash script as opposed to a something we want to actively keep in the `$PATH`

alessio commented:
@rigelrozanski the only the thing that could be automated in the Makefile is the generation of the Markdown changelog. For all the rest, I'd warmly recommend to compile it locally like all the other tools.

rigelrozanski commented:
yeah it's not a strong preference by any means, having it in cmd sounds reasonable too - I just don't want to clutter `cmd/` but maybe nbd

alessio commented:
@alexanderbez I don't mind moving it under `cmd`.

alexanderbez commented:
If its a command or ran like one (which it is -- even though it's a tool), it should live in `cmd/`.

alexanderbez commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,249 @@
+		addEntryFile(flag.Arg(1), flag.Arg(2))
+
+	case "generate":
+		version := "UNRELEASED"
```
Should this be call `PENDING CHANGELOG`?

alessio replied:
I've seen `unreleased` far more often than `pending` across various FLOSS projects. Though either doesn't bother me

alexanderbez commented on cmd/sdkch/README.md:
````diff
@@ -0,0 +1,174 @@
+
+```bash
+$ sdkch add features sdk
+[\#3813] New sdk.NewCoins safe constructor to replace bare
````
Does this open a paged reader like `git commit`? That'd be cool if it did. `-m` could take the entry message directly.

alessio replied:
That's a good idea. One could either use `-m` - and be careful about escaping characters correctly - or let the command open the editor specified in the `VISUAL`/`EDITOR` global environment variables [1]

[1] http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html

alessio replied:
Actually, I'd do it in a separate PR if we think it's really necessary. The command creates a file which you can always edit/rename later on. Is that acceptable?

alexanderbez replied:
Yes, can we do this? Many times I like to have bullet points and such and I can see that being tedious coming through just STDIN.

alexanderbez replied:
I **do** think it's necessary for a good UX flow, but it can be a separate PR sure. Do you know how to do this? I can submit a PR.

alessio replied:
Yes, I do. I see your point re: bullet points, will address asap.

rigelrozanski replied:
I think we should probably address as a part of this PR

alessio replied:
I'm on it - will update developers reference and mention of sdkch

alessio replied:
Done, please test - especially with vim as it might require a proper TTY upon execution.

rigelrozanski commented on cmd/sdkch/README.md:
````diff
@@ -0,0 +1,174 @@
+You can either drop a text file in the appropriate directory or use the `add` command:
+
+```bash
+$ sdkch add features sdk
````
This command is tremendously frustrating to use. Once you type in `sdkch add features sdk` if just waits for an endless number of new line inputs with no clear way to exit the process. Not to mention, if you mess up a line, you cannot go back to the previous line (horrid). I'd strongly recommend that an editor such as `vim` is opened _at_ the new file such that entries can be manually added in ... that or we simply do not allow for multiline input in this manner (which seems to be broken). 

Anyways yeah, this add example doesn't work

alessio replied:
Done

rigelrozanski commented on cmd/sdkch/README.md:
````diff
@@ -0,0 +1,174 @@
+```
+
+```bash
+$ sdkch generate
````
During this process (or alternatively automatically during the add process) links should be added in to any numbers provided like `\#2345` 
aka this process should convert `\#2345` -> `[\#2345](https://github.com/cosmos/cosmos-sdk/issues/2345)`

alessio replied:
We have a separate tool for that: `linkify`

alessio replied:
...but i've done it anyway

rigelrozanski commented on Makefile:
```diff
@@ -123,6 +123,7 @@ devtools-stamp: tools
 	@echo "--> Downloading linters (this may take awhile)"
 	$(GOPATH)/src/github.com/alecthomas/gometalinter/scripts/install.sh -b $(GOBIN)
 	go get github.com/tendermint/lint/golint
+	go install ./cmd/sdkch
```
we should be able to execute a workable version (maybe one that manually asks for each input such as the section and stanza, name and contents) from a command like `make cl`. A new developer should not be required to understand `sdkch` it should just be easy and intuitive when this run `make cl`

alessio replied:
I don't get it, what should `make cl` do? Compile sdkch or execute it?

alessio replied:
I don't understand this. What should `make cl` do? `sdkch` is supposed to be a developer tool, I fail to grasp the use case of running it from `Make`.

alexanderbez replied:
I have to agree here with @alessio -- there is no benefit of adding a command to make where you'll just have to end up providing all the input anyway.

alessio replied:
@rigelrozanski are you perhaps asking for some sort of interactive mode? I mean, what are the expections on new developers? I suppose it's fair to assume they would be able to launch `sdkch -help` and quickly learn its usage.

However, since all comments are addressed are three +1 have been given, can we get this over the line ASAP please? cc @cwgoes

rigelrozanski replied:
k leave for now

rigelrozanski commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,277 @@
+	switch cmd {
+
+	case "init":
+		initChangesDir()
```
I think that we should delete the command `init` altogether and just check during an add if things are initialized or not during the first add.

alessio replied:
Init is gone

rigelrozanski commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,277 @@
+		if flag.NArg() > 1 {
+			version = strings.Join(flag.Args()[1:], " ")
+		}
+		generateChangelog(version)
```
After the generate command is run, all the files are left in tact, which I think is fine if for instance the releaser messed up and wanted to re-generate the changelog information. However this indicates to me that we should also have an easy command to delete the pending changelog contents such as `sdkch clear`

alessio replied:
It's now implemented as a flag: `-prune`

rigelrozanski commented on cmd/sdkch/README.md:
```diff
@@ -0,0 +1,174 @@
+# sdkch
+Simple tool to maintain modular changelogs
```
The flow of this documentation should be restructured to move from a basic users needs (aka only needs to know the add command) to the needs of somebody making a release (needs to know generate, as well as the structure etc).

alessio replied:
Ok

alessio replied:
I've revisited README, please review

rigelrozanski commented on cmd/sdkch/README.md:
```diff
@@ -0,0 +1,174 @@
+$ sdkch add features sdk
+[\#3813] New sdk.NewCoins safe constructor to replace bare
+sdk.Coins{} declarations.
+Written to .pending/features/sdk/3813-New-sdk-NewCoins-safe-constructor-to-replace-bare
```
The message that is output should provide the full directory and a bit more information about usage maybe something like:
```
Pending changelog entry written to: 
     ~/go/src/cosmos/cosmos-sdk/.pending/features/sdk/3813-New-sdk-NewCoins-safe-constructor-to-replace-bare
to modify this entry please edit or delete the above file directly
```

alessio replied:
Done

alessio commented on cmd/sdkch/README.md:
````diff
@@ -0,0 +1,174 @@
+First, you need to initialise a new directory that will host entry files:
+
+```bash
+$ sdkch init
````
Yes

alessio commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,277 @@
+
+Commands:
+    init
+    add SECTION STANZA [MESSAGE]    Add an entry file.
```
Why? Many Unix commands help screens use all caps names for their arguments

rigelrozanski replied:
It's inconsistent with the SDK and cobra, also section and stanza's should be in square brackets because they are also arguments. if anything we could use `[optional-message]` if we wanted to differentiate

alessio replied:
Not a big deal, will amend accordingly

alessio commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,277 @@
+    add SECTION STANZA [MESSAGE]    Add an entry file.
+                                    If MESSAGE is empty, read from STDIN until it
+                                    encounters EOF.
+    generate [VERSION]              Generate a changelog in Markdown format and print
```
Could easily be done with a one liner. Don't forget that this tool will be integrated in our future automated release process, rarely a human will run `sdkch generate`

rigelrozanski replied:
yeah I agree - not a big priority - I kinda maybe like it manual a bit more

alessio commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,277 @@
+	log.Fatalf("unknown command -- '%s'\nTry '%s -help' for more information.", cmd, progName)
+}
+
+// DONTCOVER
```
It avoids this file nil coverage to affect project's coverage report

rigelrozanski replied:
oh IC ... we should add a note about this somewhere in our documentation or CONTRIBUTING.md

alessio replied:
will do

alessio commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,277 @@
+Maintain unreleased changelog entries in a modular fashion.
+
+Commands:
+    init
```
Init is unnecessary, will remove it in favour of `clear`

alessio replied:
actually, both `init` and `clear` are unnecessary. You just remove the `.pending` dir to clear pending

alessio commented:
@alexanderbez dixit:

> Also, shouldn't we update our contribution doc? 

I think we should update the PR template, there is nothing to update in CONTRIBUTING.md

alessio commented on cmd/sdkch/main.go:
```diff
@@ -0,0 +1,277 @@
+	flag.Usage = printUsage
+}
+
+func main() {
```
done

rigelrozanski replied:
this doesn't work for me? still get the same comment

rigelrozanski replied:
Oh installation isn't working...

alessio commented on cmd/sdkch/README.md:
```diff
@@ -0,0 +1,174 @@
+
+## Initialise a new entry files directory
+
+First, you need to initialise a new directory that will host entry files:
```
Once per release cycle, after cutting a new release, we should just delete the directory `.pending`

alessio commented on cmd/sdkch/README.md:
````diff
@@ -0,0 +1,174 @@
+## Generate the full changelog
+
+```bash
+find .pending -type f
````
removed

cwgoes commented:
Great tooling addition!

A few notes:
- Fails the linter
- Can `sdkch` just prepend to `CHANGELOG.md` by default?
- Can we ensure `-prune` also works when specified after the command?

alessio commented:
@cwgoes comments were addressed in person, nolint: errcheck has been reintroduced

alessio commented:
@cwgoes @rigelrozanski @alexanderbez all comments have been addressed. Please review again

alessio commented:
@rigelrozanski
1. > installation doesn't seem to be working from make devtools even after make devtools_clean

`Makefile` was buggy, fixed now:

```
alessio@bangalter:~/.../cosmos/cosmos-sdk$ which sdkch
alessio@bangalter:~/.../cosmos/cosmos-sdk$ make devtools-clean
cd /home/alessio/work/tendermint/bin && rm -f dep gometalinter statik goimports
rm devtools-stamp
alessio@bangalter:~/.../cosmos/cosmos-sdk$ make devtools
mkdir -p /home/alessio/work/tendermint/src/github.com/golang && (test ! -d /home/alessio/work/tendermint/src/github.com/golang/dep && cd /home/alessio/work/tendermint/src/github.com/golang && git clone https://github.com/golang/dep) || true &&  cd /home/alessio/work/tendermint/src/github.com/golang/dep && git fetch origin && git checkout -q 22125cfaa6ddc71e145b1535d4b7ee9744fefff2
cd /home/alessio/work/tendermint/src/github.com/golang/dep/cmd/dep && /usr/local/bin/go install
mkdir -p /home/alessio/work/tendermint/src/github.com/alecthomas && (test ! -d /home/alessio/work/tendermint/src/github.com/alecthomas/gometalinter && cd /home/alessio/work/tendermint/src/github.com/alecthomas && git clone https://github.com/alecthomas/gometalinter) || true &&  cd /home/alessio/work/tendermint/src/github.com/alecthomas/gometalinter && git fetch origin && git checkout -q v3.0.0 && cd /home/alessio/work/tendermint/src/github.com/alecthomas/gometalinter && /usr/local/bin/go install
mkdir -p /home/alessio/work/tendermint/src/github.com/rakyll && (test ! -d /home/alessio/work/tendermint/src/github.com/rakyll/statik && cd /home/alessio/work/tendermint/src/github.com/rakyll && git clone https://github.com/rakyll/statik) || true &&  cd /home/alessio/work/tendermint/src/github.com/rakyll/statik && git fetch origin && git checkout -q v0.1.5 && cd /home/alessio/work/tendermint/src/github.com/rakyll/statik && /usr/local/bin/go install
go get golang.org/x/tools/cmd/goimports
--> Downloading linters (this may take awhile)
/home/alessio/work/tendermint/src/github.com/alecthomas/gometalinter/scripts/install.sh -b /home/alessio/work/tendermint/bin
alecthomas/gometalinter info checking GitHub for latest tag
alecthomas/gometalinter info found version: 3.0.0 for v3.0.0/linux/amd64
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/gometalinter
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/gocyclo
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/nakedret
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/misspell
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/gosec
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/golint
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/ineffassign
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/goconst
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/errcheck
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/maligned
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/unconvert
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/dupl
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/structcheck
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/varcheck
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/safesql
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/deadcode
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/lll
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/goimports
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/gotype
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/staticcheck
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/interfacer
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/unparam
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/gochecknoinits
alecthomas/gometalinter info installed /home/alessio/work/tendermint/bin/gochecknoglobals
go get github.com/tendermint/lint/golint
go install ./cmd/sdkch
touch devtools-stamp
alessio@bangalter:~/.../cosmos/cosmos-sdk$ which sdkch
/home/alessio/work/tendermint/bin/sdkch
```

2. > when attempting an add I got the error sdkch: couldn't open an editor: exec: "editor": executable file not found in $PATH
I would mention in this error maybe here that the $EDITOR environment variable can be set to achieve the same thing (which is what I ended up using)

Good one. Fixed.

3. > during adding, what's with the first slash here?? shouldn't it be [\#1543] instead of \[\#1543]

`sdkch generate` does the whole thing for you. When you add a new entry just use the `#nnn` syntax, don't prepend it with slashes or square brackets whatsoever. Just write `#5643 whuats up`

4. > I'm confused as to what -prune is supposed to do (doesn't seem to do much)

Delete all files under `.pending`. It does not delete directories.

5. > #3828 (comment)

Responded inline. I don't understand what `make cl` should do, usage requirements are a bit unclear to me.

6. > #3828 (comment)

Responded inline. Fixed.

7. > need to add new instructions within .github/PULL_REQUEST_TEMPLATE.md

Done

alexanderbez commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -9,7 +9,7 @@ v    If a checkbox is n/a - please still include it but + a little note why
 - [ ] Wrote tests
 - [ ] Updated relevant documentation (`docs/`)
-- [ ] Added entries in `PENDING.md` with issue # 
+- [ ] Added a relevant changelog entry: `sdkch add`
```
```suggestion
- [ ] Added a relevant changelog entry: `sdkch add [SECTION] [STANZA] #issue ...`
```

alexanderbez commented on cmd/sdkch/README.md:
```diff
@@ -0,0 +1,51 @@
+Maintain unreleased changelog entries in a modular fashion.
+
+Commands:
+    init
```
```suggestion
```

alexanderbez commented on cmd/sdkch/README.md:
```diff
@@ -0,0 +1,51 @@
+    bugfixes                 sdk
+                      tendermint
+
+Flags:
```
Missing prune flag.

alexanderbez commented on cmd/sdkch/README.md:
````diff
@@ -0,0 +1,51 @@
+## Generate the full changelog
+
+```bash
+$ sdkch generate
````
```suggestion
$ sdkch generate v0.30.0
```

alessio commented:
I'll address your `make cl` in an ensuing PR

cwgoes commented:
(presumably we want release process integration, but that can happen separately)
